### PR TITLE
DMP-1743: Support a redis based context regisry around jwt and documentum tokens

### PR DESCRIPTION
**Pull request by mestebanez**
### JIRA link (if applicable) ###

[DMP-1743](https://tools.hmcts.net/jira/browse/DMP-1743)

### Change description ###

This ticket implements the underlying redis support for two cache implementations:-

* jwt cache - Whereby we always return a jwt token and work against that jwt for downstream darts api communication
* documentum - Whereby we translate from a documentum token to a jwt token which is then used for downstream darts communication (this is a fallback option if the above jwt implementation fails to integrate). Also, its worth mentioning that a benefit of this approach would be that the client is not subject to the timeout of the jwt token (which can in theory kick in before the time to idle). By adopting this approach the client can use the same token whilst transparently acquiring a new jwt token on the backend

**Additional Functionality**

Caches can be configured with the ability for tokens to be shared across more than one client (mid tier) request (based on the soap username/password credentials). This will reduce load on the backend Idp. 

We implement a time to idle policy whereby each  redis entry timeout expiration is reset each time it is touched expiration time.

Time to idlle is an hour by default (this achieves parity with the documentum context registry implementation)

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

**Review by hemantasharma1129 — APPROVED** [review]

File: build.gradle [hunk 1]
```diff
@@ -298,7 +298,11 @@ dependencies {
 
   implementation group: 'org.springframework.boot', name: 'spring-boot-starter-validation'
   implementation group: 'org.springframework.boot', name: 'spring-boot-starter-web-services'
+  implementation group: 'org.springframework.boot', name: 'spring-boot-starter-cache'
+  implementation group: 'org.springframework.boot', name: 'spring-boot-starter-data-redis'
   implementation group: 'org.springframework', name: 'spring-context'
+  implementation 'org.springframework.integration:spring-integration-redis:6.2.1'
+  testImplementation 'it.ozimov:embedded-redis:0.7.2'
```
Comment (richardbull) [inline build.gradle:305]: Should we not be trying to keep the dependencies together in some similar sort of style
- should these be in the `group, name, version` style?
- keep the `testImplementation` stuff together also

File: src/integrationTest/java/uk/gov/hmcts/darts/ws/ContextRegistryDocumentumToJwtServiceSharedTokenTest.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+@ActiveProfiles("int-test-documentum-jwt-token-shared")
+class ContextRegistryDocumentumToJwtServiceSharedTokenTest extends ContextRegistryParent {
+    @MockBean
+    private OauthTokenGenerator generator;
+
+    @Autowired
+    private CacheProperties properties;
+
+    @Autowired
+    private RedisTemplate<String, Object> restTemplate;
```
Comment (richardbull) [inline src/integrationTest/java/uk/gov/hmcts/darts/ws/ContextRegistryDocumentumToJwtServiceSharedTokenTest.java:25]: Should be renamed to be `redisTemplate`?

File: src/integrationTest/java/uk/gov/hmcts/darts/ws/ContextRegistryDocumentumToJwtServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+@ActiveProfiles("int-test-documentum-jwt-token")
+class ContextRegistryDocumentumToJwtServiceTest extends ContextRegistryParent {
+    @MockBean
+    private OauthTokenGenerator generator;
+
+    @Autowired
+    private CacheProperties properties;
+
+    @Autowired
+    private RedisTemplate<String, Object> restTemplate;
```
Comment (richardbull) [inline src/integrationTest/java/uk/gov/hmcts/darts/ws/ContextRegistryDocumentumToJwtServiceTest.java:25]: Ditto

File: src/integrationTest/java/uk/gov/hmcts/darts/ws/ContextRegistryJwtServiceSharedTokenTest.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+@ActiveProfiles("int-test-jwt-token-shared")
+class ContextRegistryJwtServiceSharedTokenTest extends ContextRegistryParent {
+    @MockBean
+    private OauthTokenGenerator generator;
+
+    @Autowired
+    private CacheProperties properties;
+
+    @Autowired
+    private RedisTemplate<String, Object> restTemplate;
```
Comment (richardbull) [inline src/integrationTest/java/uk/gov/hmcts/darts/ws/ContextRegistryJwtServiceSharedTokenTest.java:26]: redisTemplate

File: src/integrationTest/java/uk/gov/hmcts/darts/ws/ContextRegistryJwtServiceTest.java [hunk 5]
```diff
@@ -1,130 +1,73 @@
...
-class ContextRegistryJwtServiceTest extends IntegrationBase {
+class ContextRegistryJwtServiceTest extends ContextRegistryParent {
     @MockBean
     private OauthTokenGenerator generator;
 
+    @Autowired
+    private CacheProperties properties;
+
+    @Autowired
+    private RedisTemplate<String, Object> restTemplate;
```
Comment (richardbull) [inline src/integrationTest/java/uk/gov/hmcts/darts/ws/ContextRegistryJwtServiceTest.java:26]: redisTemplate

File: src/integrationTest/java/uk/gov/hmcts/darts/ws/ContextRegistryParent.java [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+public class ContextRegistryParent extends IntegrationBase {
+
+    void executeTestGetContextRegistryWsdl() throws Exception {
+        HttpRequest request = HttpRequest.newBuilder()
+                .uri(new URI(getGatewayUri() + "runtime/ContextRegistryService?wsdl"))
+                .timeout(Duration.ofMinutes(2))
+                .header("Content-Type", "application/json")
```
Comment (richardbull) [inline src/integrationTest/java/uk/gov/hmcts/darts/ws/ContextRegistryParent.java:33]: Is this the right header value?

File: src/main/java/uk/gov/hmcts/darts/cache/token/ServiceContextCacheValue.java [hunk 7]
```diff
@@ -1,20 +1,131 @@
...
+    public void setContextString(String context) {
+        this.contextStr = context;
+    }
+
+    @JsonIgnore
+    public ServiceContext getServiceContext() throws CacheException {
+
+        if (context == null) {
+            try {
+                JAXBContext jaxbContext = JAXBContext.newInstance(ServiceContext.class);
```
Comment (richardbull) [inline src/main/java/uk/gov/hmcts/darts/cache/token/ServiceContextCacheValue.java:68]: I think this JAXBContext.newInstance stuff should only be done in the constructor, so we would use a class field like in SoapHeaderConverter?

File: build.gradle (hunk 1 above)
Comment (mestebanez) [reply build.gradle:305]: Done

File: src/integrationTest/java/uk/gov/hmcts/darts/ws/ContextRegistryDocumentumToJwtServiceSharedTokenTest.java (hunk 2 above)
Comment (mestebanez) [reply src/integrationTest/java/uk/gov/hmcts/darts/ws/ContextRegistryDocumentumToJwtServiceSharedTokenTest.java:25]: Done. Actually wasnt being used so I have removed.

File: src/integrationTest/java/uk/gov/hmcts/darts/ws/ContextRegistryDocumentumToJwtServiceTest.java (hunk 3 above)
Comment (mestebanez) [reply src/integrationTest/java/uk/gov/hmcts/darts/ws/ContextRegistryDocumentumToJwtServiceTest.java:25]: Done

File: src/integrationTest/java/uk/gov/hmcts/darts/ws/ContextRegistryJwtServiceSharedTokenTest.java (hunk 4 above)
Comment (mestebanez) [reply src/integrationTest/java/uk/gov/hmcts/darts/ws/ContextRegistryJwtServiceSharedTokenTest.java:26]: Done

File: src/integrationTest/java/uk/gov/hmcts/darts/ws/ContextRegistryJwtServiceTest.java (hunk 5 above)
Comment (mestebanez) [reply src/integrationTest/java/uk/gov/hmcts/darts/ws/ContextRegistryJwtServiceTest.java:26]: Done

File: src/integrationTest/java/uk/gov/hmcts/darts/ws/ContextRegistryParent.java (hunk 6 above)
Comment (mestebanez) [reply src/integrationTest/java/uk/gov/hmcts/darts/ws/ContextRegistryParent.java:33]: Nope. Superflous for the GET anyhow. Removing. 

Done

File: src/main/java/uk/gov/hmcts/darts/cache/token/ServiceContextCacheValue.java (hunk 7 above)
Comment (mestebanez) [reply src/main/java/uk/gov/hmcts/darts/cache/token/ServiceContextCacheValue.java:67]: yep you are right. Done

**Pull request merged**